### DISCUSSION TITLE
Bandcamp album player

### DISCUSSION THREAD

**Discussion by joshdavenport** [johnlindquist/kit#920] [Share]
![Screenshot 2022-10-19 at 14 23 37](https://user-images.githubusercontent.com/757828/196703457-f4f02bfe-55d2-4a44-b2b1-a01060e1064c.png)

Run the script, give it a bandcamp URL (tracks also work) then as long as the artist has allowed embedding (most of them seem to, I think you have to explicitiy disable it) then you get a floating bandcamp player!

Here are some URLs to have a play with:

* https://garekdruss.bandcamp.com/album/soft-fascination-2
* https://mansurbrown.bandcamp.com/album/naqi-vol-1
* https://surprisechef.bandcamp.com/album/education-recreation
* https://pyecorneraudio.bandcamp.com/album/social-dissonance

---

[Open bandcamp-album-player in Script Kit](https://scriptkit.com/api/new?name=bandcamp-album-player&url=https://gist.githubusercontent.com/joshdavenport/184ec5c89a766f931b0a11aef2e19e46/raw/82715fca347300e40dd90a6c23e26f599c01fe1c/bandcamp-album-player.ts")

```js
// Name: Bandcamp Album Player
// Description: Stream a bandcamp album
// Author: Josh Davenport-Smith
// Twitter: @jdprts

import "@johnlindquist/kit";
import { WidgetAPI } from "@johnlindquist/kit/types/pro";
const jsdom = await npm('jsdom');

const INITIAL_PLAYER_WIDTH = 500;
const MAX_PLAYER_WIDTH = 740;
const INITIAL_PLAYER_HEIGHT = 120;

const htmlDecode = (input: string) => {
    const doc = new jsdom.JSDOM(`<!DOCTYPE html><div>${input}</div>`);
    return doc.window.document.querySelector('div').textContent;
}

const getStyles = () => `
:root {
    --dragbar-width: 30px;
}

iframe {
    position: absolute;
    left: var(--dragbar-width);
    top: 0;
    width: calc(100% - var(--dragbar-width));
    height: 100%;
    -webkit-app-region: no-drag;
}

#dragbar {
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    width: var(--dragbar-width);
    cursor: grab;
    background: #F1F1F1;
}

#dragbar:after {
    content: '';
    position: absolute;
    inset: 6px;
    background: url("data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSI1IiBoZWlnaHQ9IjUiPgo8cmVjdCB3aWR0aD0iNSIgaGVpZ2h0PSI1IiBmaWxsPSIjRjFGMUYxIj48L3JlY3Q+CjxyZWN0IHdpZHRoPSIyIiBoZWlnaHQ9IjIiIGZpbGw9IiNjY2MiPjwvcmVjdD4KPC9zdmc+");
}

#dragbar:active {
    cursor: grabbing;
}
`;

const getScripts = (albumId: number) => `
(() => {
    const url = 'https://bandcamp.com/EmbeddedPlayer/album=${albumId}/size=large/bgcol=ffffff/linkcol=0687f5/tracklist=false/artwork=small/transparent=true/';
    const iframe = document.createElement('iframe');
    iframe.setAttribute('src', url);
    iframe.setAttribute('seamless', true);
    document.body.appendChild(iframe);
})();
`;

// Obtain the album (or track) URL
const albumUrl = await arg('What is the album URL?');

// Bail if the URL is invalid
if (!albumUrl.includes('bandcamp.com/album') && !albumUrl.includes('bandcamp.com/track')) {
    throw new Error('Sorry, I don\'t recognise that URL');
}

// Give feedback that we're metadata for the album
div(md(`
Loading album...
---
Note: Some URLs will not work because of permissions set by the artist
`));

// Grab the bandcamp page for bc-page-properties which stores meta about the album
const pageContents = await fetch(albumUrl).then((res) => res.text());
const bandcampMetaParse = pageContents.match(/<meta name="bc-page-properties".*?content="([^"]+)">/);

await Promise.all([
    // Give the user a chance to read the message (some URLs will not work and there's no obious way to tell in advance)
    new Promise((resolve) => setTimeout(resolve, 2000)),
    // Also wait for the data to load, so we continue when both promises resolve
    pageContents,
])

if (!bandcampMetaParse) {
    throw new Error('Sorry, I couldn\'t find the album ID');
}

// Parse the meta we found
const bandcampMeta: { item_type?: string, item_id?: number, tralbum_page_version?: number } | undefined = JSON.parse(htmlDecode(bandcampMetaParse[1]));

if (!bandcampMeta?.item_id) {
    throw new Error('Sorry, I couldn\'t find the album ID');
}

// Create a widget with the player!
const w: WidgetAPI = await widget(
    `
<div id="dragbar"></div>
<style type="text/css">${getStyles()}</style>
<script type="text/javascript">${getScripts(bandcampMeta.item_id)}</script>
    `,
    { 
        alwaysOnTop: true, 
        width: INITIAL_PLAYER_WIDTH, 
        maxWidth: MAX_PLAYER_WIDTH,
        height: INITIAL_PLAYER_HEIGHT, 
        minHeight: INITIAL_PLAYER_HEIGHT, 
        maxHeight: INITIAL_PLAYER_HEIGHT,
        roundedCorners: false,
        thickFrame: false,
    }
);

w.setSize(INITIAL_PLAYER_WIDTH, INITIAL_PLAYER_HEIGHT);
```

**Comment by johnlindquist** [comment]
Awesome! Love everything you're sharing ❤️

Do you have any requests/feedback for widgets now that you've made a few? Always curious if there's something I can improve for you

**Reply by joshdavenport** [reply]
Thanks! If you hadn't already noticed I'm getting _a lot_ of value out of your amazing tool, it's honestly so cool.

I think the biggest pain point is management of the HTML/CSS/JS that makes up the widget itself. We (the devs building widgets) could bootstrap something or agree on some standard but it would be great to have first class support for how this ought to be organised. Even further then would be something like support for JSX & React. This could all imply a multi-file script though which has further challenges: how to move beyond a single-file script, what about version control, package delivery. Probably way too much, these scripts are really supposed to be small bits of small functionality. I wouldn't blame you at all for thinking that is way out of scope.

The API itself is plenty powerful, especially as you had the foresight to provide a way into the options passed to electron for the window, that is very nice.

**Reply by johnlindquist** [reply]
I think a "good" solution for React support is enabling the widget to load a local url then have the script start up a vite/whatever server to host it.

The React project would have to live in a separate directory, but the script could have some "clone if directory doesn't exist" logic if someone wants to share the script.


I'd probably also have to make a "useKit" hook to enable the communication between the script and the widget, but it all feels pretty straight-forward in my mind.

**Reply by joshdavenport** [reply]
That sounds great! Does that mean there would be a small bit of cold boot time for each script that needed a vite/whatever server?

Maybe this is too out there, but management of the multi-file project, and even any challenges around delivery of a complicated script and potentially multiple versions of it, could come from support for an NPM package as a script: https://gist.github.com/joshdavenport/b3f72c1f44616511934198dd54d53b08. I'm not sure if that's a wacky idea, or OTT, but it certainly seems nice superficially at least.

**Reply by johnlindquist** [reply]
Yeah, the cold-boot would only be as long as it takes vite to start up. Should be < 1 second.

I could probably make it so the widget points to a file (instead of a localhost url) and have the app manage all the vite/serve/port stuff so there's nothing to configure.

We already have "kenvs" to support multi-file sharing. Not sure if you've used them yet as they're not super well documented
